src/Language/PureScript/TypeChecker/Kinds.hs

```diff
@@ -602,15 +603,18 @@ inferDataDeclaration moduleName (ann, tyName, tyArgs, ctors) = do
           tyCtor' = foldl (\ty -> srcTypeApp ty . srcTypeVar . fst) tyCtor tyArgs'
           ctorBinders = fmap (fmap (fmap Just)) $ sigBinders <> fmap (nullSourceAnn,) tyArgs'
       for ctors $ \ctor ->
-        fmap ((ctor,) . mkForAll ctorBinders) $ inferDataConstructor tyCtor' ctor
+        fmap (mkForAll ctorBinders) <$> inferDataConstructor tyCtor' ctor
 
 inferDataConstructor
   :: forall m. (MonadError MultipleErrors m, MonadState CheckState m)
   => SourceType
   -> DataConstructorDeclaration
-  -> m SourceType
-inferDataConstructor tyCtor =
-  flip checkKind E.kindType . foldr ((E.-:>) . snd) tyCtor . dataCtorFields
+  -> m (DataConstructorDeclaration, SourceType)
+inferDataConstructor tyCtor DataConstructorDeclaration{..} = do
+  tyCtor' <- checkKind (foldr ((E.-:>) . snd) tyCtor dataCtorFields) E.kindType
+  let (_, _, tys) = unapplyTypes tyCtor'
```

Comment:
I don't think `unapplyTypes` is correct here. `unapplyTypes` deconstructs  applications, but `tyCtor'` is the signature for the data constructor. That is, given `data Foo a = Foo Int a String`, then `tyCtor'` will be `Int -> a -> String -> Foo a`. `unapplyTypes` called on this will yield an application to `->` with two arguments. Maybe we should do a traversal over `dataCtorFields`, checking each field against `E.kindType`, zip this with  dataCtorFields, and then assemble the constructor signature after.

Reply:
Of course 🤦 I‘ve checked each field independently. I attempted to build the constructor type from the checked fields types but this led to issues with rank-n kinds so I built it from the unchecked field types and then checked the whole.

Is there a way to reuse some of the work done checking the fields when building the constructor type?

Reply:
I'm not sure why you would have rank-n issues. Maybe you constructed the associativity of `->` incorrectly? Maybe something like:
```
inferDataConstructor tyCtor DataConstructorDeclaration{..} = do
  dataCtorFields' <- traverse (traverse (flip checkKind E.kindType)) dataCtorFields
  dataCtor <- flip (foldr ((E.-:>) . snd)) dataCtorFields' =<< checkKind tyCtor E.kindType
  pure (DataConstructorDeclaration { dataCtorFields = dataCtorFields, .. }, dataCtor)
```

Reply:
Oh I just forgot to check the constructor return type :sweat_smile: I applied your suggestion!